### PR TITLE
blog post 'Natural Language Processing for Icelandic with PyPy: A Case Study'

### PR DESCRIPTION
**Pull request by cfbolz**
draft blog post by Vilhjálmur Þorsteinsson (@vthorsteinsson)

### CODE REVIEW

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+Icelandic language. The project focuses on collecting data and
+developing open source software for a range of core applications, such
+as tokenization, vocabulary lookup, n-gram statistics, part-of-speech
+tagging, named entity recognition, spelling and grammar checking, neural
+language models and speech processing.
+
+------------
+
+My name is Vilhjálmur Þorsteinsson, and I’m the founder and CEO of a
+`software startup <https://mideind.is/english.html>`__ in Reykjavík,
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:49]:
```suggestion
software startup `Miðeind <https://mideind.is/english.html>`__ in Reykjavík,
```

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+for the government program, the requirements were for a major, well
+supported, vendor-and-OS-agnostic FOSS platform with a large and diverse
+community, including in the NLP space. The decision to select Python as
+a foundational language for the project was a relatively easy one. That
+said, there was a bit of trepidation around the well known fact that
+CPython can be slow for inner-core tasks, such as tokenization and
+parsing, that can see heavy workloads in production.
+
+I first became aware of PyPy in early 2016 when I was developing `a
+crossword game <https://github.com/mideind/Netskrafl>`__ in Python 2.7
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:67]:
```suggestion
I first became aware of PyPy in early 2016 when I was developing a
crossword game `Neskrafl <https://github.com/mideind/Netskrafl>`__ in Python 2.7
```

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+In NLP, speed and memory requirements can be quite important for
+software usability. On the other hand, NLP logic and algorithms are
+often complex and challenging to program, so programmer productivity and
+code clarity are also critical success factors. A pragmatic approach
+balances these factors, avoids premature optimization and seeks a
+careful compromise between maximal run-time efficiency and minimal
+programming and maintenance effort.
+
+Turning to our use cases, our `Icelandic text
+tokenizer <https://github.com/mideind/Tokenizer>`__ is fairly light,
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:89]:
```suggestion
Turning to our use cases, our Icelandic text
tokenizer `"tokenizer" <https://github.com/mideind/Tokenizer>`__ is fairly light,
```

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+runs tight loops and performs a large number of small, repetitive
+operations. It runs very well on PyPy’s JIT and has not required further
+optimization.
+
+Our `Icelandic parser <https://github.com/mideind/GreynirPackage>`__ is,
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:94]:
```suggestion
Our Icelandic parser `GreynirPackage <https://github.com/mideind/GreynirPackage>`__
(known on PyPI as `reynir <https://pypi.org/project/reynir/>`__) is,
```

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+although we had to figure out a few `magic incantations in our build
+module <https://github.com/mideind/GreynirPackage/blob/master/src/reynir/eparser_build.py>`__
+to make it compile smoothly during setup from source on Windows and
+MacOS in addition to Linux. Of course, we build binary PyPy and CPython
+wheels for the most common targets so most users don’t have to worry
+about setup requirements.
+
+With the positive experience from the parser project, we proceeded to
+take a similar approach for two other core NLP packages: our `compressed
+vocabulary package <https://github.com/mideind/BinPackage>`__ and our
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:132]:
```suggestion
take a similar approach for two other core NLP packages: our compressed
vocabulary package `BinPackage <https://github.com/mideind/BinPackage>`__ 
(known on PyPI as `islenska <https://pypi.org/project/islenska/>`__ and our
```

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+`trigrams database package <https://github.com/mideind/Icegrams>`__.
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:133]:
```suggestion
trigrams database package `Icegrams <https://github.com/mideind/Icegrams>`__.
```

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+workloads, it  became apparent that it needed to be faster still. We
+then proceeded to convert the innermost Earley parsing loop from Python
+to `tight
+C++ <https://github.com/mideind/GreynirPackage/blob/master/src/reynir/eparser.cpp>`__
+and to call it from PyPy via
+`CFFI <https://cffi.readthedocs.io/en/latest/>`__, with callbacks for
+token-terminal matching functions (“business logic”) that remained on
+the Python side. This made the parser much faster (on the order of 100x
+faster than the original on CPython) and quick enough for our production
+use cases.
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:120]: Question: did PyPy still give a speed boost over CPython even after the transition to CFFI?

**Review by mattip — COMMENTED** [review]
I think using the package names in the post will make them more discoverable.

**Comment by mattip** [comment]
Cool!
[here](https://deploy-preview-50--keen-mestorf-442210.netlify.app/posts/2022/02/nlp-icelandic-case-study.html) is the draft render of the post.

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 8]
```diff
@@ -0,0 +1,168 @@
+.. title: Natural Language Processing for Icelandic with PyPy: A Case Study
+.. slug: nlp-icelandic-case-study
+.. date: 2022-02-06 15:00:00 UTC
+.. tags: casestudy
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:4]: So far this is the only "casestudy". Some candidates (links to this PR's rendered html):
- [matrix multiplication](https://deploy-preview-50--keen-mestorf-442210.netlify.app/posts/2018/06/repeating-matrix-multiplication-8641748742577945875.html)
- [a calculator in RPython](https://deploy-preview-50--keen-mestorf-442210.netlify.app/posts/2018/11/guest-post-implementing-calculator-repl-6271483514675006846.html)
- [making code 80x faster](https://deploy-preview-50--keen-mestorf-442210.netlify.app/posts/2017/10/how-to-make-your-code-80-times-faster-1424098117108093942.html)

Should we add more tags as part of this PR?

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 1 above)
Comment (vthorsteinsson) [reply posts/2022/02/nlp-icelandic-pypy.rst:49]: Agreed

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 2 above)
Comment (vthorsteinsson) [reply posts/2022/02/nlp-icelandic-pypy.rst:67]: Neskrafl -> Netskrafl (missing 't')

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 3 above)
Comment (vthorsteinsson) [reply posts/2022/02/nlp-icelandic-pypy.rst:89]: "Tokenizer" with a capital T?

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 4 above)
Comment (vthorsteinsson) [reply posts/2022/02/nlp-icelandic-pypy.rst:94]: GreynirPackage -> Greynir - that is the name of the parser toolset as such, but for legacy reasons the website repo is called Greynir and the Python package is GreynirPackage. A bit confusing, I know.

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 7 above)
Comment (vthorsteinsson) [reply posts/2022/02/nlp-icelandic-pypy.rst:120]: Yes, definitely - there is a lot of other performance-sensitive logic, especially the parse tree reducer which remains in Python and is still 3-4x faster on PyPy than CPython.

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 5 above)
Comment (vthorsteinsson) [reply posts/2022/02/nlp-icelandic-pypy.rst:132]: Agreed

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 6 above)
Comment (vthorsteinsson) [reply posts/2022/02/nlp-icelandic-pypy.rst:133]: Agreed

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 2 above)
Comment (mattip) [reply posts/2022/02/nlp-icelandic-pypy.rst:67]: fixed

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 3 above)
Comment (mattip) [reply posts/2022/02/nlp-icelandic-pypy.rst:89]: On PyPI it is with a lower-case t, on github it is capital T. I think `pip install Tokenizer` will work so let's go with the capital T.

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 4 above)
Comment (mattip) [reply posts/2022/02/nlp-icelandic-pypy.rst:94]: As far as I can tell it is called `reynir` on PyPI, no?

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 7 above)
Comment (mattip) [reply posts/2022/02/nlp-icelandic-pypy.rst:120]: So can we add here "Even after moving much of the heavy processing to C++ and using CFFI, PyPy still gives a signficant speed boost over CPython."?

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 4 above)
Comment (vthorsteinsson) [reply posts/2022/02/nlp-icelandic-pypy.rst:94]: Yes, adding even more to the confusion ;-)

**Comment by vthorsteinsson** [comment]
Looks good to me!

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 7 above)
Comment (vthorsteinsson) [reply posts/2022/02/nlp-icelandic-pypy.rst:120]: Yes, that is entirely correct.

**Comment by cfbolz** [comment]
Cool, thanks Matti and @vthorsteinsson! I'll work in Matti's suggestions and then post in a bit.

**Comment by cfbolz** [comment]
Ah, Matti did, I missed that 😅

File: posts/2022/02/nlp-icelandic-pypy.rst (hunk 4 above)
Comment (mattip) [reply posts/2022/02/nlp-icelandic-pypy.rst:94]: OK, changing as requested

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+workloads, it  became apparent that it needed to be faster still. We
+then proceeded to convert the innermost Earley parsing loop from Python
+to `tight
+C++ <https://github.com/mideind/GreynirPackage/blob/master/src/reynir/eparser.cpp>`__
+and to call it from PyPy via
+`CFFI <https://cffi.readthedocs.io/en/latest/>`__, with callbacks for
+token-terminal matching functions (“business logic”) that remained on
+the Python side. This made the parser much faster (on the order of 100x
+faster than the original on CPython) and quick enough for our production
+use cases.
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:121]:
```suggestion
use cases. Even after moving much of the heavy processing to C++ and using CFFI, PyPy still gives a significant speed boost over CPython.
```

**Comment by mattip** [comment]
here is the latest version https://deploy-preview-50--keen-mestorf-442210.netlify.app/posts/2022/02/nlp-icelandic-case-study.html

**Comment by vthorsteinsson** [comment]
One typo left: There is a missing right parenthesis after `"(known on PyPI as [islenska](https://pypi.org/project/islenska/)"`.

Apart from that, good to go from my side!

File: posts/2022/02/nlp-icelandic-pypy.rst [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+module <https://github.com/mideind/GreynirPackage/blob/master/src/reynir/eparser_build.py>`__
+to make it compile smoothly during setup from source on Windows and
+MacOS in addition to Linux. Of course, we build binary PyPy and CPython
+wheels for the most common targets so most users don’t have to worry
+about setup requirements.
+
+With the positive experience from the parser project, we proceeded to
+take a similar approach for two other core NLP packages: our compressed
+vocabulary package `BinPackage <https://github.com/mideind/BinPackage>`__ 
+(known on PyPI as `islenska <https://pypi.org/project/islenska/>`__ and our
```
Comment (mattip) [inline posts/2022/02/nlp-icelandic-pypy.rst:134]:
```suggestion
(known on PyPI as `islenska <https://pypi.org/project/islenska/>`__) and our
```

**Comment by mattip** [comment]
> There is a missing right parenthesis after ...

Fixed

**Comment by cfbolz** [comment]
thanks @vthorsteinsson and @mattip this is really a great post, I am merging

**Pull request merged**